### PR TITLE
Support hex, octal, and binary numbers

### LINKED ISSUE #245: Support hexadecimal inputs
A separate issue from #42 which is asking about bitwise operations, I'd love for the parser just to be able accept hexadecimal integers with 0x notation. I commonly use Insect to calculate between units of bytes, which I sometimes get in hex. It would be awesome if I didn't have to convert the value to decimal using some other tool before pasting it into Insect.

Getting *output* in hex would also be awesome, and *ideally* Insect would support arbitrary bases for inputs and outputs, but in this issue I'm only formally requesting support for hexadecimal inputs.

### PR DESCRIPTION
**Pull request by mhmdanas**
Fixes #245.

I'm really, really ashamed of the terrible way I implemented this, but I honestly could not think of a way to do it that wouldn't be nearly as bad. Suggestions for alternate implementations or ways to improve the existing one are very welcome.

TODO:
- [x] Write tests

### CODE REVIEW

File: src/Insect/Parser.purs [hunk 1]
```diff
@@ -127,9 +117,28 @@ number = do
...
+      intPart ← (fromCharArray <<< fromFoldable) <$> some digit
       pure $ singleton (codePointFromChar sign) <> intPart
 
+number ∷ P Decimal
+number = do
+  numberPrefix /\ digit' /\ expSymbol ←
+    option ("" /\ digit /\ "e") $ char '0'
+      *> ((char 'x' <|> char 'X') $> ("0x" /\ hexDigit /\ "p")
+      <|> (char 'o' <|> char 'O') $> ("0o" /\ octDigit /\ "p")
+      <|> (char 'b' <|> char 'B') $> ("0b" /\ (oneOf ['0', '1'] <?> "binary digit") /\ "p"))
```
Comment (mhmdanas) [inline src/Insect/Parser.purs:129]: <sub>kill me, what have I done</sub>

**Comment by sharkdp** [comment]
> I'm really, really ashamed of the terrible way I implemented this, but I honestly could not think of a way to do it that wouldn't be nearly as bad.

I think it's perfectly fine as long as we add enough tests to make sure that everything is alright. Having a good test coverage also allows us to refactor with confidence later, if needed.

File: test/Main.purs [hunk 2]
```diff
@@ -152,6 +152,54 @@ main = runTest do
...
+         "0xC.C + 0o15.1 - 0b10.01010p3"
+
+       shouldParseAs (Expression (scalar 481.65625)) "0x1E1.a8"
+
+       -- Ensure the parser doesn't greedily consume "p".
+       allParseAs (Expression (BinOp Mul (scalar 49.0) (Variable "p")))
+         [ "0x31p"
+         , "0o61p"
+         , "0b110001p"
+         ]
```
Comment (mhmdanas) [inline test/Main.purs:193]: Should this part perhaps be in the "Parser - Identifiers" suite instead?

File: docs/features.md [hunk 3]
```diff
@@ -26,6 +26,14 @@ Documentation
...
     * **Exponential notation**: `6.022e23`.
 
+    * **Hexadecimal, octal and binary number input**:
+
+      ```
+      0x2.F
+      0o5p3
+      0b1011
+      ```
+
```
Comment (mhmdanas) [inline docs/features.md:36]: How does this look?

File: test/Main.purs [hunk 4]
```diff
@@ -152,6 +152,48 @@ main = runTest do
...
+       shouldParseAs (Expression (scalar 481.65625)) "0x1E1.a8"
+
+       shouldFail "0xG"
+       shouldFail "0oG"
+       shouldFail "0oA"
+       shouldFail "0b2"
+       shouldFail "0be"
+       shouldFail "0x1p+"
+       shouldFail "0x1p-"
+       shouldFail "0x5.."
```
Comment (sharkdp) [inline test/Main.purs:195]: :+1: :+1:

**Review by sharkdp — APPROVED** [review]
Apart from the minor docu thing, this looks great. Thank you

File: docs/features.md (hunk 3 above)
Comment (mhmdanas) [reply docs/features.md]: Done, thank you for the feedback.

**Pull request merged**